### DISCUSSION TITLE
Timeline Support

### DISCUSSION THREAD

**Discussion by wojiuxuihuan** [yasirkula/UnityBezierSolution#23] [Ideas]
Hi! I have been looking for Timeline WayPath at AssetStore for ages, but none of them reach our goal, until I found this solution, it's wonderful! 
It is widely known that developing with timeline-clip is pain in the ass, and it might take you lots of time, I know it's not fair to waste your time for these feature.
Actually I had made my own Timeline plugin [EventPlayer](https://assetstore.unity.com/packages/tools/visual-scripting/event-player-116731) which will send clip data (e.g. Time, Percent, Duration) and is easy to extend without unnecessary coding. So I wonder, can I add timeline support for your plugin?

I have test the following code and it works fine, is it ok for you to add the following method(or similar) for BezierWalker/ParticlesFollowBezier so I can invoke it manually? (Or If you are not going to change the interface in the up coming release, I can write these thing in my plugin  (^_^). 

        public void Evaluate(float normalizedT, bool isForward)
        {
            isGoingForward = isForward;
            float _normalizedT = highQuality ? spline.evenlySpacedPoints.GetNormalizedTAtPercentage(normalizedT) : normalizedT;//Smoothing

            //Position
            transform.position = spline.GetPoint(_normalizedT);

            //Rotation
            if (lookAt == LookAtMode.Forward)
            {
                BezierSpline.Segment segment = spline.GetSegmentAt(_normalizedT);
                Quaternion targetRotation;
                if (isForward)
                    targetRotation = Quaternion.LookRotation(segment.GetTangent(), segment.GetNormal());
                else
                    targetRotation = Quaternion.LookRotation(-segment.GetTangent(), segment.GetNormal());

                transform.rotation = targetRotation;
            }
            else if (lookAt == LookAtMode.SplineExtraData)
                transform.rotation = spline.GetExtraData(_normalizedT, extraDataLerpAsQuaternionFunction);
        }

**Comment by yasirkula** [comment]
I can't think of many uses for this function apart from Timeline support, so I'd prefer not adding it to the plugin. I'd recommend you to create an extension function in your plugin and call it instead:

```csharp
public static void Evaluate( this BezierWalkerWithTime walker, float normalizedT, bool isForward )
{
	float _normalizedT = walker.highQuality ? walker.spline.evenlySpacedPoints.GetNormalizedTAtPercentage( normalizedT ) : normalizedT;

	//Position
	walker.transform.position = walker.spline.GetPoint( _normalizedT );

	//Rotation
	if( walker.lookAt == LookAtMode.Forward )
	{
		BezierSpline.Segment segment = walker.spline.GetSegmentAt( _normalizedT );
		Quaternion targetRotation;
		if( isForward )
			targetRotation = Quaternion.LookRotation( segment.GetTangent(), segment.GetNormal() );
		else
			targetRotation = Quaternion.LookRotation( -segment.GetTangent(), segment.GetNormal() );

		walker.transform.rotation = targetRotation;
	}
	else if( walker.lookAt == LookAtMode.SplineExtraData )
		walker.transform.rotation = walker.spline.GetExtraData( _normalizedT, walker.extraDataLerpAsQuaternionFunction );
}
```

As for ParticlesFollowBezier, Relaxed mode probably won't work but Strict mode may work out of the box. You need to first somehow evaluate the ParticleSystem itself at that time, then evaluate the ParticlesFollowBezier script.

**Reply by wojiuxuihuan** [reply]
Thank's for the advice! Because the extraDataLerpAsQuaternionFunction is still protected, I am gonna try it another way by inherit the BezierWalker, if you have spare time to set it to public that will be great! The new update may release in this week, thanks again!

**Reply by yasirkula** [reply]
You're right, I didn't notice that issue. I may make the function public static in a future update but at the moment, I'd recommend you to copy&paste these code to your own script: https://github.com/yasirkula/UnityBezierSolution/blob/218365fb0e833f6d162a84a4a7f73408b68957ff/Plugins/BezierSolution/Utilities/BezierWalker.cs#L16-L21

**Comment by wojiuxuihuan** [comment]
Cool! I have update the Plugin, and it might help others to create TimelinePath Animation! Thanks again!
Here's the video in case someone interest: [Timeline BezierSolution Support]( https://youtu.be/Lb3xB9wVDW4)

**Reply by yasirkula** [reply]
Looks great! Thank you for sharing 😃
